### ISSUE TITLE
LLVM commit d1ad006a8f64bdc disables CONFIG_EFI_STUB

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1725]
After https://github.com/llvm/llvm-project/commit/d1ad006a8f64bdc17f618deffa9e7c91d82c444d, I noticed that `CONFIG_EFI_STUB` can no longer be selected because `-mabi=ms` is not a valid flag for x86:

```
config EFI_STUB
        bool "EFI stub support"
        depends on EFI
        depends on $(cc-option,-mabi=ms) || X86_32
        select RELOCATABLE
        help
          This kernel feature allows a bzImage to be loaded directly
          by EFI firmware without the use of a bootloader.

          See Documentation/admin-guide/efi-stub.rst for more information.
```

```
$ clang-16 -mabi=ms -Werror -c -x c -o /dev/null /dev/null
clang-16: error: argument unused during compilation: '-mabi=ms' [-Werror,-Wunused-command-line-argument]

$ clang-14 -mabi=ms -Werror -c -x c -o /dev/null /dev/null
```

As far as I can tell, this `cc-option` check was added with commit [8f24f8c2fc82](https://git.kernel.org/linus/8f24f8c2fc82f701866419dcb594e2cc1d3f46ba) ("efi/libstub: Annotate firmware routines as __efiapi") to ensure `__attribute__((ms_abi))` was able to be used but it seems like that is always available for clang, as we have seen no warnings or build errors and it appears to be supported in [`getCallingConventionForDecl()`](https://github.com/llvm/llvm-project/blob/e549ac072b55f6d29d01a31cddf246dcd1b353db/clang/lib/CodeGen/CGCall.cpp#L242-L243).

@ardbiesheuvel Is there any reason that this diff might not work?

```diff
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index f9920f1341c8..06055b64f51f 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1953,10 +1953,13 @@ config EFI
          resultant kernel should continue to boot on existing non-EFI
          platforms.

+config CC_HAS_MS_ABI_FN_ATTR
+       def_bool $(success,echo '__attribute__((ms_abi)) int x(void);' | $(CC) $(CLANG_FLAGS) -x c - -c -o /dev/null -Werror)
+
 config EFI_STUB
        bool "EFI stub support"
        depends on EFI
-       depends on $(cc-option,-mabi=ms) || X86_32
+       depends on CC_HAS_MS_ABI_FN_ATTR || X86_32
        select RELOCATABLE
        help
          This kernel feature allows a bzImage to be loaded directly
```

**Comment by nathanchance** [comment]
Ha, I cannot actually boot my machines without `CONFIG_EFI_STUB` through `systemd-boot`, apparently. I get the following all red error message on my monitor hooked up to one of my test systems:

```
Error loading \vmlinuz-linux-next-llvm: Unsupported.
Failed to execute Arch Linux (-next + LLVM) (\vmlinuz-linux-next-llvm): Unsuppor
```

That change gets everything working for me again but I am unsure if there are other implications so I'll wait for Ard's feedback before I send it.

**Comment by ardbiesheuvel** [comment]
Let's just drop that `depends on` line entirely - I don't think we need it given the oldest supported GCC version is 5.1 these days.

**Comment by nathanchance** [comment]
Sounds good, I will send out a patch tomorrow, as I am running out of time today.

[On Compiler Explorer](https://godbolt.org/z/f1ooGd1PM), I can see `__attribute__((ms_abi))` is supported back to at least GCC 4.4.7 and `-mabi=ms` is supported by at least GCC 4.5.3. I see 8f24f8c2fc82f701866419dcb594e2cc1d3f46ba landed in 5.6, which has a documented minimum supported version of GCC 4.9 (although it was not enforced until 5.12). Would it be acceptable to mark the upstream patch for stable (to account for those who might be on 5.10 and 5.15 but using a newer toolchain like ChromeOS) or should we be conservative and do something like

```diff
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 57f5e881791a..4b48b99a682f 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1926,7 +1926,7 @@ config EFI
 config EFI_STUB
        bool "EFI stub support"
        depends on EFI && !X86_USE_3DNOW
-       depends on $(cc-option,-mabi=ms) || X86_32
+       depends on $(cc-option,-mabi=ms) || CC_IS_CLANG || X86_32
        select RELOCATABLE
        help
          This kernel feature allows a bzImage to be loaded directly
```

just for stable?

**Comment by ardbiesheuvel** [comment]
Nah just rip it out and backport it to v5.10+. We only use the attribute not the command line option, so worst case, people [supposedly] building v5.10 with GCC 4.3 or older will have to disable CONFIG_EFI_STUB by hand in their .config after upgrading to a new stable release.

**Comment by nathanchance** [comment]
Patch submitted, thanks for the input Ard!

https://lore.kernel.org/20220929152010.835906-1-nathan@kernel.org/

**Comment by nathanchance** [comment]
Patch accepted: https://git.kernel.org/tip/33806e7cb8d50379f55c3e8f335e91e1b359dc7b

**Comment by nathanchance** [comment]
Landed in 6.1-rc2: https://git.kernel.org/linus/33806e7cb8d50379f55c3e8f335e91e1b359dc7b

It is cc'd for stable so it will get there eventually.

**Comment by nickdesaulniers** [comment]
stable backports: https://lore.kernel.org/stable/Y1lhm3mNdI0PFbLe@dev-arch.thelio-3990X/
